### PR TITLE
mixed layer documentation

### PR DESCRIPTION
**Pull request by matthewjhenry**

### CODE REVIEW

**Comment by penmaher** [comment]
Hi Matthew. Thanks for getting this done. I am happy to review this but the earliest I can start is the end of next week. Sorry for the delay.

File: docs/source/modules/mixedlayer.rst [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+| albedo            | Surface albedo                      | None                               |
++-------------------+-------------------------------------+------------------------------------+
+
+
+Relevant modules and subroutines
+--------------------------------
+.. List the names of relevant modules, subroutines, functions, etc.
+.. You can add also code snippets, using Sphinx code formatting
+
+The mixed layer code is located in: src/atmos_spectral/driver/solo/mixed_layer.F90. The name of this file reflects the fact that the code determines the properties of the single layer (either a slab ocean model
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:252]: The file locations here (4 total) don't jump out at me. Suggest using ``link`` notation to make them red.

File: docs/source/modules/mixedlayer.rst [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+--------------------------------
+.. List the names of relevant modules, subroutines, functions, etc.
+.. You can add also code snippets, using Sphinx code formatting
+
+The mixed layer code is located in: src/atmos_spectral/driver/solo/mixed_layer.F90. The name of this file reflects the fact that the code determines the properties of the single layer (either a slab ocean model 
+or prescribed SST)  below the air-sea interface.
+
+The mixed layer ocean is initialised and called by:  src/atmos_spectral/driver/solo/idealized_moist_phys.F90
+
+Relevant routines which (which have not already been listed in SST options above) are called by the mixed layer ocean are:
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:257]: I don't think this clause is needed. Easier to read without it. If you like bullet points here I suggest using the same for the .F90 files above too.

File: docs/source/modules/mixedlayer.rst [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
++-------------------------------+-------------------------------------------------------------------------------------------------------------+----------------------+
+|``add_latent_heat_flux_anom``  | Flag to add an anomalous latent heat flux                                                                   | ``False``            |
++-------------------------------+-------------------------------------------------------------------------------------------------------------+----------------------+
+
+Diagnostics
+-------------------
++-------------------+-------------------------------------+------------------------------------+
+| Name              | Description                         | Units                              |
++===================+=====================================+====================================+
+| t_surf            | Surface temperature                 | :math:`\text{K}\`                  |
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:231]: t_surf and delta_t_surf have not rendered properly. I can't see any bug here but perhaps it is the last '\'.

File: docs/source/modules/mixedlayer.rst [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
++-------------------+-------------------------------------+------------------------------------+
+| delta_t_surf      | Surface temperature change          | :math:`\text{K}\`                  |
++-------------------+-------------------------------------+------------------------------------+
+| flux_t            | Surface sensible heat flux          | :math:`\text{W}\,\text{m}^{-2}`    |
++-------------------+-------------------------------------+------------------------------------+
+| flux_lhe          | Surface latent heat flux            | :math:`\text{W}\,\text{m}^{-2}`    |
++-------------------+-------------------------------------+------------------------------------+
+| flux_oceanq       | Ocean heat flux                     | :math:`\text{W}\,\text{m}^{-2}`    |
++-------------------+-------------------------------------+------------------------------------+
+| ice_conc          | Sea ice concentration               | None                               |
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:241]: Perhaps rather than saying none in these two, say Unitless with a range of 0-1 or similar?

File: docs/source/modules/mixedlayer.rst [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
++-------------------+-------------------------------------+------------------------------------+
+| albedo            | Surface albedo                      | None                               |
++-------------------+-------------------------------------+------------------------------------+
+
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:245]: Can you add land_sea_heat_capacity please?

File: docs/source/modules/mixedlayer.rst [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
++-------------------+----------------------------------------------------------------+---------+
+|``do_qflux``       | Switch to calculate time-independent q-flux.                   |``False``|
++-------------------+----------------------------------------------------------------+---------+
+|``qflux_amp``      | Amplitude of time-independent q-flux if ``do_qflux`` is True.  | ``0.0`` |
++-------------------+----------------------------------------------------------------+---------+
+|``qflux_width``    | Width of time-independent q-flux if ``do_qflux`` is True.      | ``16.0``|
++-------------------+----------------------------------------------------------------+---------+
+
++----------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+
+| Option               | Summary                                                                                                                                                                    |Default          |
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:109]: Why are these as 2 tables? I think line wrapping would look better than a scrolling bar. What is your perspective?

File: docs/source/modules/mixedlayer.rst [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
++------------------------------+---------------------------------------------------------------------------------------------------------+----------+
+|``elandlat``                  | Vector determining higher bounds of latitudes for land masses.                                          | ``-1``   |
++------------------------------+---------------------------------------------------------------------------------------------------------+----------+
+|``land_h_capacity_prefactor`` | Factor by which to multiply ocean heat capacity to get land heat capacity if ``input`` option is used.  | ``1.0``  |
++------------------------------+---------------------------------------------------------------------------------------------------------+----------+
+|``land_albedo_prefactor``     | Factor by which to multiply ocean albedo to get land albedo if ``input`` option is used.                | ``1.0``  |
++------------------------------+---------------------------------------------------------------------------------------------------------+----------+
+
+Other options (temporary title)
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:195]: Title update needed.You missed the following options:
trop_depth, trop_cap_limit, heat_cap_limit, np_cap_factor, flux_lhe_anom_file_name, flux_lhe_anom_field_name

File: docs/source/modules/mixedlayer.rst [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+Summary
+----------------------
+This module updates the sea surface temperature (SST, :math:`T_s`). 
+
+SST boundary condition options
+-----------------------
+The SST options are:
+    - an input SST file (``do_sc_sst`` namelist option)
+    - a slab ocean aquaplanet analytic form (``prescribe_initial_dist`` namelist option)
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:12]: Can you add a reference here please? I assume one of Frierson's papers but check the form in the paper to make sure.

Comment (penmaher) [reply docs/source/modules/mixedlayer.rst:12]: Sorry don't put the ref here but in the section below.

File: docs/source/modules/mixedlayer.rst [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+    - an AquaPlanet Experiment protocol (APE) analytic form (``do_ape_sst`` namelist option)
+    - an implicit timestep (``do_calc_eff_heat_cap`` namelist option)
+    - an optional Q-flux (this is not controlled by a flag but rather an input file that is generated offline).
+
+Each of these options are discussed below.
+
+Input SST file
+^^^^^^^^^^^^^^^
+Set ``do_sc_sst`` to True if you want to specify the SST field. The SST field will be read in from an input filename using the ``sst_file`` variable.
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:21]: mentioned somewhere in here the file needs to be NetCDF

File: docs/source/modules/mixedlayer.rst [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+^^^^^^^^^^^^^^^
+Set ``do_sc_sst`` to True if you want to specify the SST field. The SST field will be read in from an input filename using the ``sst_file`` variable. 
+Using an input SST field is useful, for example,  when you want to add a temperature anomaly. The file is read in during the initialisation (i.e. within the call to mixed_layer_init from within idealized_moist_phys).
+The input file can be time independent (i.e. no diurnal or seasonal cycle or any changes in the SST from one time step to another) or vary with time. More information can be found in the Diagnostics section below.
+
+Slab ocean 
+^^^^^^^^^^^^^^^
+During the initialisation, if there is no restart file to open, the surface temperature is set to the prescribed initial distribution (``prescribe_initial_dist = True``)
+
+:math:`T_s = T_{surf} -\frac{1}{3}*dT*\left(3\sin(\lambda)^2-1\right)`,
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:29]: the * looks strange here. Suggest removing.

File: docs/source/modules/mixedlayer.rst [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+
+where the default values we use in the trip test for the Frierson test case (exp/test_cases/frierson/frierson_test_case.py) are: :math:`T_{surf} = 285 K` and :math:`dT = 40 K`.
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:31]: use `` link`` notation to make it red

File: docs/source/modules/mixedlayer.rst [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+APE aquaplanet (analytic form)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The prescribed SST for the APE aquaplanet protocol is given by
+
+:math:`T_s = 27 \left( 1 - \sin^2\left( \frac{3}{2} \lambda \right) \right)`,
+
+between 60N-60S, equation 1 of Neele and Hoskins 2004 [NealeHoskins2004]_, and 0 deg C poleward of 60N/S.
+
+
+Implicit timestep
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:44]: Add you add a general english sentence her eon what implicit time stepping is please.

File: docs/source/modules/mixedlayer.rst [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+
+eff_heat_capacity * dTs/dt = - corrected_flux.
+
+Additional Q-flux
+^^^^^^^^^^^^^^^^^^^^
+The slab ocean model only communicates between grid-boxes in the vertical (i.e. air-sea exchange) but does not represent any horizontal transport (i.e. no north-south or east-west communications between grid cells). 
+An idealised horizontal transport can be included using an ocean heat flux (q-flux). Atmospheric heat transport is more realistic with an ocean heat transport.
+
+To create a qflux file, run a prescribed run (i.e. a control) using either observations, AMIP or similar. The prescribed SST field and the surface fluxes (from this control run)
+are then used to compute the Q-flux by running the offline python script: src/extra/python/scripts/calculate_qflux/calculate_qflux.py. This script produces a .nc file which is then passed into the model
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:72]: ``link`` notation please

File: docs/source/modules/mixedlayer.rst [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+| Option            | Summary                                                    |Default  |
++===================+============================================================+=========+
+|``evaporation``    |Switch for surface evaporation.                             |``True`` |
++-------------------+------------------------------------------------------------+---------+
+|``depth``          |Mixed layer depth.                                          | ``40.0``|
++-------------------+------------------------------------------------------------+---------+
+
+Q-flux options
+^^^^^^^^^^^^^^^^^^^^
+If ``do_qflux`` is True, use ``qflux_amp`` and ``qflux_width`` to calculate a time-independent surface q-flux.
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:94]: Could you add a sentence above in the q-flux description to include how the qflux_amp and qflux_width work?

File: docs/source/modules/mixedlayer.rst [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+land_sea_heat_capacity * dTs/dt = - corrected_flux - t_surf_dependence * dTs
+
+The change in surface temperature in this timestep (dTs/dt) times the heat capacity of the surface is equal to the net flux into the surface at the given timestep (- corrected_flux) plus the change in the net flux 
+into the surface caused by the change in surface temperature (- t_surf_dependence * dTs).
+
+This is simplified by defining eff_heat_capacity as land_sea_heat_capacity + t_surf_dependence * dt, and then Ts is updated using
+
+eff_heat_capacity * dTs/dt = - corrected_flux.
+
+Additional Q-flux
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:66]: Perhaps 'Optional' rather than 'Additional'?

File: docs/source/modules/mixedlayer.rst (hunk 14 above)
Comment (penmaher) [reply docs/source/modules/mixedlayer.rst:94]: Okay I get why there are two tables now. Could you separate the text above into two sections corresponding to each table please? Might make it more obvious.

File: docs/source/modules/mixedlayer.rst [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
++-------------------+-----------------------------------------------------------------------------+---------+
+|``albedo_cntr``    | Parameter that sets surface albedo for albedo choice 5.                     | ``45.0``|
++-------------------+-----------------------------------------------------------------------------+---------+
+|``albedo_wdth``    | Parameter that sets surface albedo for albedo choice 5.                     | ``10``  |
++-------------------+-----------------------------------------------------------------------------+---------+
+
+Land options
+^^^^^^^^^^^^^^^^
+
+There are X options for setting up the land, determined by the value of ``land_option``.
```
Comment (penmaher) [inline docs/source/modules/mixedlayer.rst:168]: X=4?

**Review by penmaher — CHANGES_REQUESTED** [review]
Hi Mat, This is looking great. I have done my first read though which was to make sure nothing was missing and things make sense. Which on a whole they do :-) Lay look great and is logical for this doc. I plan to do a second read through to check the science bits. I will do that this afternoon and add any relevant comments on the PR front page. Penny

**Comment by penmaher** [comment]
Science things to check. Line ref are for mixed_layer.F90

- L339:  The T_s and T_surf notation in documentation is a little confusing. Suggest using T_C or another form (unless it is stated T_surf in a paper somewhere in which case it if probably fine).

- In the surface albedo options section, the notation in 1-5 if hard to read/follow. There are some notation errors here too (eg the end of item 4). Could you work on making this easier to understand please?

-  The implicit time stepping equation (land_sea_heat_capacity * dTs/dt = - corrected_flux - t_surf_dependence * dTs) I can't find it in this form in the code.  L721 is similar. Can you check this form please? In general I found this section on time stepping a little hard to follow (notation is a primary reason why I found it hard to follow - maybe a different appearance for the variable names might help readability?)
- description in table for tconst and delta_T mention a formulas above, suggest you say in what section

- the albedo section has 2 sets of variables with the same description. Can you make it clearer how they are different?

- in the 'others' section. Perhaps a 'sea ice' section might help. Also do_ape_sst, sst_file and some others here  would work in the section "Initialize surface temperature". Perhaps anything else that does not fit could be described  as 'Optional extras'?

Okay that is all I see for now. Over to you @matthewjhenry

**Comment by matthewjhenry** [comment]
Hi Penny, 

Thanks for all the comments. Here are a few responses.

I know they are in the code, but are the "trop_depth, trop_cap_limit, heat_cap_limit, np_cap_factor, flux_lhe_anom_file_name, flux_lhe_anom_field_name" options still used? Somehow I remember reading they weren't hence weren't added to the doc, but maybe I am wrong on this.

>     * L339:  The T_s and T_surf notation in documentation is a little confusing. Suggest using T_C or another form (unless it is stated T_surf in a paper somewhere in which case it if probably fine).

L339 refers to mixed_layer.F90 code? Confused.

>     * The implicit time stepping equation (land_sea_heat_capacity * dTs/dt = - corrected_flux - t_surf_dependence * dTs) I can't find it in this form in the code.  L721 is similar. Can you check this form please? 

It's not used, the expression under it is used : eff_heat_capacity * dTs/dt = - corrected_flux (L. 713 in mixed_layer.F90). I don't think adding mixed_layer.F90 line numbers makes sense, as it may change a lot in the future.

>     * the albedo section has 2 sets of variables with the same description. Can you make it clearer how they are different?

I think it's maybe clearer now. It just depends on what albedo_choice one chooses.

Let me know if you have any other comments, also don't hesitate to upload changes directly.

-Matthew

File: docs/source/modules/mixedlayer.rst (hunk 7 above)
Comment (penmaher) [reply docs/source/modules/mixedlayer.rst]: Did you decide not to include these options or was this missed?

File: docs/source/modules/mixedlayer.rst (hunk 10 above)
Comment (penmaher) [reply docs/source/modules/mixedlayer.rst]: The math in this section "APE aquaplanet (analytic form)" is not rendering correctly

File: docs/source/modules/mixedlayer.rst (hunk 7 above)
Comment (penmaher) [reply docs/source/modules/mixedlayer.rst]: Sorry only now saw your response from 9 Dec. No problem to not include them.

**Comment by penmaher** [comment]
Sorry for the delay in responding. 

If I said to include Line numbers then that is a terrible suggestion - I can't remember making it and can't easily see it now. I think I might have only been referring to line numbers to provide context for what I was saying. 

Only a few minor things to change:
1. I think the definition of Ts is not as obvious as it could be. Suggest replacing "(SST, Ts)" with "(SST) noted as $T_s$ below." I know it is more wordy but I think it is clearer.
2. I mented above but will say here too in case it is not seen. The formula for T_s in "APE aquaplanet (analytic form)" has not rendered correctly.

I think once those two small things are done we are ready to pass to Stephen for review.

**Comment by matthewjhenry** [comment]
Thank you for the corrections @penmaher !  :)
I did the suggested edits along with a few more minor corrections.

Over to you @sit23 for the review, and then to @RuthG for quality control.

**Review by penmaher — APPROVED** [review]

**Comment by penmaher** [comment]
Approval from me for this. @sit23  over to you.

**Comment by sit23** [comment]
OK - I realistically won't get to this for a week or two. Sorry about that. Will do it as soon as I can.

**Comment by penmaher** [comment]
I am going to make some additions to this page, after the group discussion, to give some more context to the q-flux part. Once that is done I will commit the changes to this PR and then pass over to @sit23

**Comment by penmaher** [comment]
@matthewjhenry can you check out my suggested changes and pass onto @sit23 if you are happy with them.

@sit23, in your review, can you critique the new section on q-fluxes, specifically step 2. When I have created q-fluxes in the past I have not used your script which is references in the doc. Could you check this makes sense and it suitable please?

**Comment by matthewjhenry** [comment]
Checked your last commit @penmaher, looks a lot clearer, I have nothing to add. Thanks for fixing all the lower-case q's! :)

Over to you, @sit23.

**Comment by RuthG** [comment]
Looks great, thanks Matthew!  
I'm open to disagreement, but I have some suggestions to restructure things to hopefully make it a little clearer. 
I've tried to suggest copy-paste-able text inserts to make this easier because personally I really appreciate these from coauthors etc., but no obligation to actually copy-paste them if you think they're nonsense!

l8-15: It might be clearer to reframe this introduction to overview the broad mixed layer options in the module, rather than trying to outline the namelist parameters. e.g.
```
The SST options are:
    - prescribe SST from an input file (``do_sc_sst`` namelist option)
    - prescribe SST to follow an AquaPlanet Experiment protocol (APE) analytic form (``do_ape_sst`` namelist option)
    - calculate SST based on the surface fluxes and mixed layer depth of a **slab ocean**, with the option of including a Q-flux (either analytic or read from a file)

Each of these options are discussed below, followed by an outline of the implicit time-stepping process used when SST is not prescribed. Note that only the final case will generate a closed surface energy budget.
```

I'd then suggest reordering the subsections to match the list i.e.:
- Input SST file
- APE aquaplanet (analytic SST)
- Slab ocean
- Implicit timestepping procedure


l22:  mixed_layer_init and  idealized_moist_phys would be good in `` ``s

l27: Perhaps basic but for novice users you could briefly explain what a slab ocean is here e.g.
```
To allow the SST to evolve based on the surface fluxes, the atmosphere can be coupled to a slab ocean, whose depth is specified by the namelist parameter ``depth``. This also allows a closed surface energy budget, useful for e.g. simulations with increased greenhouse gases.

In this case, during the initialisation....
```

I'd also suggest folding the Q-flux outline (and perhaps the implicit timestepping as well actually) into this section, and then it's very clear it's only relevant to 'coupled' slab ocean runs. 


l76: This is true where the Q-flux aims to mimic an input SST field. Some users may also want to use an idealised Q-flux with a different form to those included in the model. So maybe this bit could be generalised a little e.g.

```
Isca is able to calculate an analytic Q-flux, appropriate to an aquaplanet, following Merlis et al 2013 [MerlisEtAl2013] if ``do_qflux`` is True. Additionally, an analytic warmpool can be added  by setting ``do_warmpool``.  The warmpool structure is set in ``Isca/src/atmos_param/qflux/qflux.f90``. Alternatively an arbitrary Q-flux may be read from a file if ``load_qflux`` is True. 

In the case where both a specific SST distribution (e.g. AMIP climatology) and closed surface energy budget are desired, a Q-flux input file can be generated by running a control experiment with the prescribed SST, creating a NetCDF Q-flux file offline and then passing this file to the model via the python interface run script.

1. Run....
```

Otherwise looks good to me! The hardcoded land and albedo options look intrinsically confusing
